sklearn/metrics/cluster/supervised.py

```diff
@@ -188,33 +204,49 @@ def adjusted_rand_score(labels_true, labels_pred, max_n_classes=5000):
     adjusted_mutual_info_score: Adjusted Mutual Information
 
     """
-    labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
-    n_samples = labels_true.shape[0]
-    classes = np.unique(labels_true)
-    clusters = np.unique(labels_pred)
+    if contingency is None:
+        labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
+        n_samples = labels_true.shape[0]
+        n_classes = np.unique(labels_true).shape[0]
+        n_clusters = np.unique(labels_pred).shape[0]
+    elif isinstance(contingency, _data_matrix):  # scipy.sparse.data._data_matrix
+        n_samples = contingency.nnz
+        n_classes, n_clusters = contingency.shape
+    else:
+        raise ValueError("'contingency' must be a sparse matrix or None")
+
     # Special limit cases: no clustering since the data is not split;
     # or trivial clustering where each document is assigned a unique cluster.
     # These are perfect matches hence return 1.0.
-    if (classes.shape[0] == clusters.shape[0] == 1 or
-            classes.shape[0] == clusters.shape[0] == 0 or
-            classes.shape[0] == clusters.shape[0] == len(labels_true)):
+    if (n_classes == n_clusters == 1 or
+                    n_classes == n_clusters == 0 or
+                    n_classes == n_clusters == n_samples):
         return 1.0
 
-    contingency = contingency_matrix(labels_true, labels_pred,
-                                     max_n_classes=max_n_classes)
+    # Compute contingency matrix if we weren't given it
+    if contingency is None:
+        contingency = contingency_matrix(labels_true, labels_pred, max_n_classes=max_n_classes)
 
     # Compute the ARI using the contingency data
-    sum_comb_c = sum(comb2(n_c) for n_c in contingency.sum(axis=1))
-    sum_comb_k = sum(comb2(n_k) for n_k in contingency.sum(axis=0))
+    if isinstance(contingency, np.ndarray):
+        # For an array
+        sum_comb_c = sum(comb2(n_c) for n_c in contingency.sum(axis=1))
+        sum_comb_k = sum(comb2(n_k) for n_k in contingency.sum(axis=0))
+        sum_comb = sum(comb2(n_ij) for n_ij in contingency.flatten())
+    elif isinstance(contingency, _data_matrix):
+        # For a sparse matrix
+        sum_comb_c = sum(comb2(n_c) for n_c in np.array(contingency.sum(axis=1)))
+        sum_comb_k = sum(comb2(n_k) for n_k in np.array(contingency.sum(axis=0)).T)
+        sum_comb = sum(comb2(n_ij) for n_ij in find(contingency)[2])
```

Comment:
If it's a `_data_matrix` surely `contingency.data.flatten()` will do without the more expensive `find` operation?